freqtrade/strategy/hyper.py

```diff
@@ -0,0 +1,93 @@
+"""
+IHyperStrategy interface, hyperoptable Parameter class.
+This module defines a base class for auto-hyperoptable strategies.
+"""
+from abc import ABC
+from typing import Union, List, Iterator, Tuple
+
+from skopt.space import Integer, Real, Categorical
+
+from freqtrade.strategy.interface import IStrategy
+
+
+class Parameter(object):
+    """
+    Defines a parameter that can be optimized by hyperopt.
+    """
+    default: Union[int, float, str, bool]
+    space: List[Union[int, float, str, bool]]
+    category: str
+
+    def __init__(self, *, space: List[Union[int, float, str, bool]], default: Union[int, float, str, bool] = None,
+                 category: str = None, **kwargs):
+        """
+        Initialize hyperopt-optimizable parameter.
+        :param space: Optimization space. [min, max] for ints and floats or a list of strings for categorial parameters.
+        :param default: A default value. Required for ints and floats, optional for categorial parameters (first item
+         from the space will be used). Type of default value determines skopt space used for optimization.
+        :param category: A parameter category. Can be 'buy' or 'sell'. This parameter is optional if parameter field
+         name is prefixed with 'buy_' or 'sell_'.
+        :param kwargs: Extra parameters to skopt.space.(Integer|Real|Categorical).
+        """
+        assert 'name' not in kwargs, 'Name is determined by parameter field name and can not be specified manually.'
+        self.value = default
+        self.space = space
+        self.category = category
+        self._space_params = kwargs
+        if default is None:
+            assert len(space) > 0
+            self.value = space[0]
+
+    def get_space(self, name: str) -> Union[Integer, Real, Categorical, None]:
+        """
+        Create skopt optimization space.
+        :param name: A name of parameter field.
+        :return: skopt space of this parameter, or None if parameter is not optimizable (i.e. space is set to None)
+        """
+        if not self.space:
+            return None
+        if isinstance(self.value, int):
+            assert len(self.space) == 2
+            return Integer(*self.space, name=name, **self._space_params)
+        if isinstance(self.value, float):
+            assert len(self.space) == 2
+            return Real(*self.space, name=name, **self._space_params)
+
+        assert len(self.space) > 0
+        return Categorical(self.space, name=name, **self._space_params)
```

Comment:
I could not agree more! Separate parameter classes is a much better idea. Especially since it would allow categorial parameter to have no default (because `default='foo', space=['foo', 'bar']` isnt really nice), while mandating default value for float/int params as i do not think there can be a sensible default for those.